pkg/ccl/crosscluster/logical/logical_replication_writer_processor.go

```diff
@@ -674,6 +680,30 @@ func (lrw *logicalReplicationWriterProcessor) flushBuffer(
 		return a.KeyValue.Value.Timestamp.Compare(b.KeyValue.Value.Timestamp)
 	})
 
+	// If the seen map is nil or has hit 2M items, reset it.
+	if lrw.seenKeys == nil || len(lrw.seenKeys) > 2<<20 {
+		lrw.seenKeys = make(map[uint64]int64, 2<<20)
+	}
+
+	h := fnv.New64a()
+	logged := false
+	for i := range kvs {
+		h.Reset()
+		_, _ = h.Write(kvs[i].KeyValue.Key)
+		hashed := h.Sum64() + uint64(kvs[i].KeyValue.Value.Timestamp.WallTime)
+		c := lrw.seenKeys[hashed]
+		lrw.seenKeys[hashed] = c + 1
+
+		if c > 0 {
+			lrw.dupeCount++
+			if !logged && lrw.seenEvery.ShouldLog() {
+				logged = true // don't check ShouldLog again for rest of loop.
+				log.Infof(ctx, "duplicate delivery of key %s@%d (%d prior times); %d total recent dupes.",
```

Comment:
did you intend to reset `dupeCount`, given the phrase "recent dupes". If not, could be nice to include dupeCount in the vtable, as a lower bound on duplicates seen in the proc.

Reply:
> and only those that it has seen since it last reset its map

but you would need to reset dupecount when you reset the hash then, right?

Reply:
sg on vtable exclusion. I think the only change to make is to reset `dupeCount` when you reset `seenKeys`